emacs_templates/emacs/elisp/lang-web.el

```diff
@@ -1,32 +1,94 @@
-  (add-to-list 'auto-mode-alist '("\\.html?\\'" . web-mode))
-  (add-to-list 'auto-mode-alist '("\\.phtml\\'" . web-mode))
-  (add-to-list 'auto-mode-alist '("\\.tpl\\.php\\'" . web-mode))
-  (add-to-list 'auto-mode-alist '("\\.[agj]sp\\'" . web-mode))
-  (add-to-list 'auto-mode-alist '("\\.as[cp]x\\'" . web-mode))
-  (add-to-list 'auto-mode-alist '("\\.erb\\'" . web-mode))
-  (add-to-list 'auto-mode-alist '("\\.mustache\\'" . web-mode))
-  (add-to-list 'auto-mode-alist '("\\.djhtml\\'" . web-mode))
-  (add-to-list 'auto-mode-alist '("\\.jsx$" . web-mode))
-  (add-hook 'web-mode-hook 'jsx-flycheck))
-
-
-(defadvice web-mode-highlight-part (around tweak-jsx activate)
-  (if (equal web-mode-content-type "jsx")
-      (let ((web-mode-enable-part-face nil))
-        ad-do-it)
-    ad-do-it))
-
-(defun jsx-flycheck ()
-  (when (equal web-mode-content-type "jsx")
-              ;; enable flycheck
-              (flycheck-select-checker 'jsxhint-checker)
-              (flycheck-mode)))
+  (add-hook 'web-mode-hook 'jsx-flycheck)
+
+  ;; highlight enclosing tags of the element under cursor
+  (setq web-mode-enable-current-element-highlight t)
+
+  (defadvice web-mode-highlight-part (around tweak-jsx activate)
+    (if (equal web-mode-content-type "jsx")
+        (let ((web-mode-enable-part-face nil))
+          ad-do-it)
+      ad-do-it))
+
+  (defun jsx-flycheck ()
+    (when (equal web-mode-content-type "jsx")
+      ;; enable flycheck
+      (flycheck-select-checker 'jsxhint-checker)
+      (flycheck-mode)))
+
+  ;; editing enhancements for web-mode
+  ;; https://github.com/jtkDvlp/web-mode-edit-element
+  (use-package web-mode-edit-element
+    :config (add-hook 'web-mode-hook 'web-mode-edit-element-minor-mode))
+
+  ;; snippets for HTML
+  ;; https://github.com/smihica/emmet-mode
+  (use-package emmet-mode
+    :init (setq emmet-move-cursor-between-quotes t) ;; default nil
+    :diminish (emmet-mode . " e")
+    :bind (("C-c ]" . emmet-next-edit-point)
+           ("C-c [" . emmet-prev-edit-point)))
+  (add-hook 'web-mode-hook 'emmet-mode)
+
+  (defun my-web-mode-hook ()
+    "Hook for `web-mode' config for company-backends."
+    (set (make-local-variable 'company-backends)
+         '((company-tern company-css company-web-html company-files))))
+  (add-hook 'web-mode-hook 'my-web-mode-hook)
+
+  ;; Enable JavaScript completion between <script>...</script> etc.
+  (defadvice company-tern (before web-mode-set-up-ac-sources activate)
+    "Set `tern-mode' based on current language before running company-tern."
+    (message "advice")
+    (if (equal major-mode 'web-mode)
+	(let ((web-mode-cur-language
+	       (web-mode-language-at-pos)))
+	  (if (or (string= web-mode-cur-language "javascript")
+		  (string= web-mode-cur-language "jsx"))
+	      (unless tern-mode (tern-mode))
+	    (if tern-mode (tern-mode -1))))))
+  (add-hook 'web-mode-hook 'company-mode)
+
+  ;; to get completion data for angularJS
+  (use-package ac-html-angular :defer t)
```

Comment:
This and the one below are `autocomplete` dependent.. and i don't really like mixing up ac & company together. So i would exclude these two packages